### ISSUE TITLE
JPG (and maybe geotiff) cropping needs fixing

### ISSUE THREAD

**Issue by jywarren** [publiclab/mapknitter-exporter-sinatra#32]
This could be at `mapknitter-exporter` level, but the final jpgs are not showing all images; it seems to be cropped to the dimensions of one of the original jpgs?

It's from: https://mapknitter.org/maps/irish-uk-border-mapping/warpables.json has ~30 images of ~2mb each!

Screenshot:

![image](https://user-images.githubusercontent.com/24359/57201059-955e3200-6f61-11e9-89d6-dd0637770b34.png)

Full image:

https://00e9e64bac16bf5f16185332ccc0c49307b4036f26402a8c29-apidata.googleusercontent.com/download/storage/v1/b/mapknitter-exports-warps/o/1557075547%2F1557075547.jpg?qk=AD5uMEsZqJDXS5Ibvf31kqPs_8sZ8qdwok4VEMYNkBxdw2UJB3zjYe8AQiij5vbuF7iBfIxOFcZbl5_OgggPxhtV6rSvQesUjAwh5TjleoRHAE3sr50TRDMSapC-gF19Aa3aZp9kcC3Tg5Nzccq_JwasHHnz2NCtLj9gUPzW6JsJRasWrcVOXrdEMezIOkN6vwSsBHrBK0U4I4mDu7NzYkYEwbzbQr5uxtF-vT6BxQMaHQXVRt9XVgHIWSzkvTsWUWkSAMVqo1CprtXQOTOJdtCnK-0csbBdYsWoSkHo4f-yhOl2Ra2Z3gF-O2sOZKWi1MDaA7WnhbC-W7-RlUw-vtqQzOKjLXipnyoWm8bK5IfhT_QFezSD77bEDBwzhw7lP1_-_gwcWQPCPZ5jYGsA8Quir85DLENjT7V0DdjGYOBiiI5h9tLhGAH6168VSOjwWy26ZhjokEJMBDdUdKYRjVFgRDRKJODAd-QQpE2tXNgK1-X70mM1PsN-acsUEKYlQQJFhXn4Hrfp5Miruhst-zW8ofR1QgjD0v9Ut_ZAvhBaE85qaXJB7M3KlyYZh4AJY1qVEPVKSNpHR4PQ5PlGJi1dsAe1lSBkE_JkAYq5MrUyhOjJmmPSwOz5cny8BOIFNcVpGsJbE1xX-HtAOJa9W_JuYKIoaxsO09ml7p0exDQu19zNnU3qrgZXDGNMI8VfG7SgsB8F-x36bpijNTxbHt1rAtfIoWiFdGyEm1Bs3zW6mq14Qiizn2MJRPmgocDeh5Sy9QlUim4aZdXRhaY-8mJAIUjOeeJl68ThWfA1hastAM1P66rNnTI

**Comment by jywarren** [comment]
OK, cool - in local testing, i found that the individual warped images (in the `ceres--2` map are OK:

![image](https://user-images.githubusercontent.com/24359/57953759-03d6b480-78bf-11e9-85d3-41b81024e4e0.png)

However the combined TIFF is not:

![image](https://user-images.githubusercontent.com/24359/57953841-40a2ab80-78bf-11e9-9fca-68804bc77b39.png)


and so the final JPG is not:

![image](https://user-images.githubusercontent.com/24359/57953775-105b0d00-78bf-11e9-8126-ce4e16bb287b.png)

**Comment by jywarren** [comment]
So, i suspect `self.generate_composite_tiff`

**Comment by jywarren** [comment]
Going to try console outputting all the lats/lons that we find.

**Comment by jywarren** [comment]
OK, here is the log output: https://gist.github.com/jywarren/ea07a6b3fbe3e5b329a06ef409f805c5

**Comment by jywarren** [comment]
So, extents are being set on line 277:

```
gdalwarp = "gdalwarp -s_srs EPSG:3857 -te #{minlon} #{minlat} #{maxlon} #{maxlat} #{directory}#{wid}-geo.tif #{directory}#{id}-geo.tif"                          
```

**Comment by jywarren** [comment]
OK, i found the minlat and maxlat values seemed to be switched. Trying to reverse those...

**Comment by jywarren** [comment]
aha yes, maybe it's reading the values as strings? Not respecting negative values? It had been saying: 

`Creating output file that is 1908P x -1115L.`

**Comment by jywarren** [comment]
```
minlat -37.7650239004, minlon 144.9828654528, maxlat -37.7668304285, maxlon 144.9859553576
gdalwarp -s_srs EPSG:3857 -te 144.9828654528 -37.7650239004 144.9859553576 -37.7668304285 public/warps/ceres2/w306187-geo.tif public/warps/ceres2/ceres2-geo.tif
Creating output file that is 1908P x -1115L.
ERROR 1: Attempt to create 1908x-1115 dataset is illegal,sizes must be larger than zero.
gdalwarp public/warps/ceres2/w306188-geo.tif public/warps/ceres2/ceres2-geo.tif
Creating output file that is 986P x 780L.
Processing input file public/warps/ceres2/w306188-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Warning 1: Lost metadata writing to GeoTIFF ... too large to fit in tag.
gdalwarp public/warps/ceres2/w306189-geo.tif public/warps/ceres2/ceres2-geo.tif
Processing input file public/warps/ceres2/w306189-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Warning 1: Lost metadata writing to GeoTIFF ... too large to fit in tag.
identify -quiet -format '%b,%w,%h' public/warps/ceres2/ceres2-geo.tif
3083340B
986
780
> generating tiles
```

**Comment by jywarren** [comment]
Yes -- so, it fails on the first image, then it defaults to the dimensions of the second image!

**Comment by jywarren** [comment]
My `.to_f` seems to have helped:

```
minlat -37.7668304285, minlon 144.9828654528, maxlat -37.7650239004, maxlon 144.9859553576
gdalwarp -s_srs EPSG:3857 -te 144.9828654528 -37.7668304285 144.9859553576 -37.7650239004 public/warps/ceres3/w306187-geo.tif public/warps/ceres3/ceres3-geo.tif
Creating output file that is 1908P x 1116L.
Processing input file public/warps/ceres3/w306187-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Warning 1: Lost metadata writing to GeoTIFF ... too large to fit in tag.
gdalwarp public/warps/ceres3/w306188-geo.tif public/warps/ceres3/ceres3-geo.tif
Processing input file public/warps/ceres3/w306188-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Warning 1: Lost metadata writing to GeoTIFF ... too large to fit in tag.
gdalwarp public/warps/ceres3/w306189-geo.tif public/warps/ceres3/ceres3-geo.tif
Processing input file public/warps/ceres3/w306189-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Warning 1: Lost metadata writing to GeoTIFF ... too large to fit in tag.
identify -quiet -format '%b,%w,%h' public/warps/ceres3/ceres3-geo.tif
8545255B
1908
1116
```

Yes! But, i think it ran out of memory?

**Comment by jywarren** [comment]
Fix at https://github.com/publiclab/mapknitter-exporter/pull/24

**Comment by jywarren** [comment]
Let's release this and run it in the production environment. I wonder if this addresses #34 as well if it failed to use the first image.

**Comment by jywarren** [comment]
Published the gem! Rebuilding the sinatra image.

**Comment by jywarren** [comment]
Once this builds, we should test output again with http://export.mapknitter.org/export?url=https://mapknitter.org/maps/ceres--2/warpables.json&scale=30

**Comment by jywarren** [comment]
![1558151339](https://user-images.githubusercontent.com/24359/57964170-bed27280-78fe-11e9-9eeb-83163f8f0770.jpg)

Hmm. Better? Not sure.

**Comment by jywarren** [comment]
I think not! Is it running on the 1.0.4 gem? We should output the gem version in the Json.

**Comment by jywarren** [comment]
Hmm, I'm seeing a 404 there. Let's wait until Monday!

**Comment by jywarren** [comment]
Hi @icarito - is this running on the `v1.0.4` gem? thanks!

**Comment by jywarren** [comment]
Comparing to 
![image](https://user-images.githubusercontent.com/24359/58189067-c344b580-7c87-11e9-86a0-485e33bfa77c.png)

**Comment by jywarren** [comment]
Oops, i hadn't updated the Gemfile.lock in this repo. Just did it: https://github.com/publiclab/mapknitter-exporter-sinatra/commit/38491dec2969012d3a93162b6eb419d2d24ce889

**Comment by jywarren** [comment]
Actually interesting, the URL for the JPG is now: https://mapknitter-exports-warps.storage.googleapis.com/1558539174/1558539174.jpg

which may mean we could update our status.json files to point directly at the container.

**Comment by jywarren** [comment]
Noting https://github.com/jywarren/web/blob/master/containers.md#how-to-deploy-and-update as well

**Comment by jywarren** [comment]
Hmm, same thing. And noticing that the image is actually reversed, lending support to the theory that it's using wrong coords and producing a negative value in the gdalwarp command, and failing subsequent steps, as identified above:

```
gdalwarp -s_srs EPSG:3857 -te 144.9828654528 -37.7650239004 144.9859553576 -37.7668304285 public/warps/ceres2/w306187-geo.tif public/warps/ceres2/ceres2-geo.tif
Creating output file that is 1908P x -1115L.
```

![image](https://user-images.githubusercontent.com/24359/58201895-3f4cf680-7ca4-11e9-84dd-4d8aed93f349.png)

Not sure why my changes to that (that seemed to work locally) aren't working here.

**Comment by jywarren** [comment]
![1558564995](https://user-images.githubusercontent.com/24359/58213908-dd9b8500-7cc1-11e9-90c2-497ff915c2a7.jpg)

Ok! Progress!!!

**Comment by jywarren** [comment]
Fixed! see #23
